packages/lodestar/src/sync/reqResp/reqResp.ts

```diff
@@ -97,7 +98,25 @@ export class SyncReqResp implements ISyncReqResp {
     } catch (e) {
       this.network.reqResp.sendResponse(id, e, null);
     }
-    // TODO handle incorrect forkVersion or disjoint finalizedCheckpoint
+    if (await this.shouldDisconnectOnHello(peerInfo, request)) {
+      this.network.reqResp.goodbye(peerInfo, 0n);
+    }
+  }
+
+  public async shouldDisconnectOnHello(peerInfo: PeerInfo, request: Hello): Promise<boolean> {
+    const headBlock = await this.db.block.getChainHead();
+    if (!headBlock) {
+      return true;
+    }
+    const state = await this.db.state.get(headBlock.stateRoot);
+    if (!state.fork.currentVersion.equals(request.headForkVersion)) {
+      return true;
+    }
+    if (state.finalizedCheckpoint.epoch >= request.finalizedEpoch &&
+       !request.finalizedRoot.equals(getBlockRoot(this.config, state, request.finalizedEpoch))) {
```

Comment:
if I'm not mistaken `getBlockRoot` only contains root for recent slots. I think it's safer to fetch block from database by request.finalizedEpoch start slot.

You should try to avoid fetching entire state here as that call isn't cheap, especially considering that this method will be called on each peer connect.

Reply:
I replaced getBlockRoot by getting block root from database as you said.

Regarding the state, I still need it for other calculations.